Draft: Multipath Will Fix This (#367)
n0-computer/iroh.computer: src/app/blog/multipath-will-fix-this/page.mdx

import { BlogPostLayout } from '@/components/BlogPostLayout'
import {ThemeImage} from '@/components/ThemeImage'

export const post = {
  draft: false,
  author: 'Floris Bruynooghe',
  date: '2025-07-29',

>   title: 'Multipath Will Fix This',

Author: So this title is basically an in-joke of the team.  Which isn't even explained in the post.  While that's a great working title, and is probably fine as a URL slug, I'm not sure this works as an actual title.

Simple.

<Note>
This post is generally going to simplify the world a lot.
Of course there are many more network situations other than two endpoints both connected to the internet via a NAT router.
And iroh has to work with all of them.
But you would get bored reading this and I would get lost writing it.
So I'm keeping this narrative simple.
</Note>

# Relay Servers

An iroh relay server is a classical piece of server software,
running in a datacenter.
It exists even though we want p2p connections,
because in today's internet we can not have direct connections without holepunching.
And you can not have holepunching without being able to coordinate.
Thus, the relay server.

Because we would like this relay server to essentially *always* work,
it uses the most common protocol on the internet:
HTTP1.1 inside a TLS stream.
Endpoints establish an entirely normal HTTPS connection to the relay server and then upgrades it to a WebSocket connection.[^websocket]
This works even in many places where the TLS connection is Machine-In-The-Middled by inserting new "trusted" root certs because of "security".

> As long as an endpoint keeps this WebSocket connection open it can use the relay server.

Author: going to leave this

  send "ping" datagrams to each candidate address of the peer.
  Both at the same time.
  
- If a "ping" datagram is received,
  respond with "yay, we holepunched!".
  Typically this will be only on 1 IP path out of all the candidates.
  Or maybe more and more these days it'll succeed for both an IPv4 and an IPv6 path.
  
If you followed carefully you'll have counted 3 special messages that need to be sent to the peer endpoint:

1. IP address candidates.  These are sent via the relay server.

2. Pings.  These are sent on the non-relayed IP paths.

3. Pongs.  These are also sent on the non-relayed IP paths.

They need to be sent as UDP datagrams.
Over the same *paths* as the QUIC datagrams are also being sent:
the **relay path** and any **direct paths**.

# Multiplexing UDP datagrams

Iroh stands on the shoulders of giants,
and it looked carefully at ZeroTier and Tailscale.
In particular it borrowed a lot from the DERP design from Tailscale.
From the above holepunching description we get two kinds of packets:

- Application payload.
  For iroh these are QUIC datagrams.
- Holepunching datagrams.

Both these need to be sent and received from the same socket,

> because holepunching a differet socket than your application data uses is not that helpful.

Author: Going to leave this.  I think the former fits the tone better.  Yes, I'm assuming the reader understands "socket" which is perhaps a bit lower-level than the general tone.  But "port" is on the same level-ish and even more only-vaguely-correct.

So when an iroh endpoint receives a packet it needs to first figure out which kind of packet this is:
a QUIC datagram, or a holepunching datagram?

> If it is a QUIC packet it is passed onto the QUIC stack.[^quicstack]

Reviewer: I'd *really* like the article to explain that we *need to do address translation in iroh* because QUIC doesn't support multipath yet! And this is where we'd say this.

Without explaining this, I think an astute reader will be left wondering "well, but you seem to be able to do multipath with implementing QUIC multipath, right? You send on relay or direct already?".

If it is a holepunching datagram it needs to be handed by iroh itself,
by a component we call the *magic socket*.
This is done using the "QUIC bit",
a bit in the UDP datagram defined as always set to 1 in QUIC version 1.[^greasing]

[^quicstack]: iroh uses [Quinn] for the QUIC stack, an excellent project.

[Quinn]: https://crates.io/crates/quinn

[^greasing]: Since QUIC has released RFC 9287 which advocates "greasing" this bit:
    effectively toggling it randomly.
    This is an attempt to stop middleboxes from ossifying the protocol by starting to recognise this bit.
    Iroh not being able to grease this bit right now is not ideal either.

> # IP Congestion Control
> 
> This system works great and is what powers iroh today.
> However, it also has its limitations.
> One interesting aspect of the internet is *congestion control*.

Reviewer: ```suggestion
# Shortcomings of the Current Approach

This system works great and is what powers iroh today.
However, it also has its limitations.

## Congestion Control

One interesting aspect of the internet is *congestion control*.
```

And maybe another subheading "## Address Translation to simulate Multipath" that explains the problems with hiding the multi-path-ness from the QUIC networking library via address translation: The fact that we don't do client migration on the QUIC stack, we don't have different CID per path we send on, as well as the fact that this is the source of all the issues with congestion control (since the QUIC stack just thinks it's the same path all the time).

Author: Oh yeah, I didn't explain why I ignored your comment on this so far.  I don't like the "negative" aspect of the "shortcomings" heading.  Though I do agree that I wasn't quite sure what heading to use, and I kind of retrofitted it after seeing the entire thing is basically talking about congestion control.

But I agree that this section needs more work to sort these issues out.

Reviewer: We could call it "Room for improvement" :D

Or perhaps matching the actual text, sth like "Limitations of the Current Approach"?

Idk, it feels honest to say it has shortcomings/limitations.

import { BlogPostLayout } from '@/components/BlogPostLayout'
import {ThemeImage} from '@/components/ThemeImage'

export const post = {
  draft: false,
  author: 'Floris Bruynooghe',

>   date: '2025-07-29',

Reviewer: Let's remember to fix the date, should it not get published today.

  title: 'Multipath Will Fix This',
  description:

>     "A lightweight rpc crate for iroh protocols",

Reviewer: The description needs to be updated.

}

export const metadata = {
    title: post.title,
    description: post.description,
    openGraph: {
      title: post.title,
      description: post.description,
      images: [{
        url: `/api/og?title=Blog&subtitle=${post.title}`,
        width: 1200,
        height: 630,
        alt: post.title,
        type: 'image/png',
      }],
      type: 'article'
    }
}

export default (props) => <BlogPostLayout article={post} {...props} />

Iroh is a library to establish direct peer-to-peer QUIC connections.

> This mean iroh does NAT traversal, colloquially known as holepunching.

Reviewer: A link for what NAT traversal is (or maybe just a NAT), might be nifty.

Author: Two candidate links:
- https://en.wikipedia.org/wiki/NAT_traversal
- https://tailscale.com/blog/how-nat-traversal-works

I'm kind of leaning towards the tailscale link, if we are happy to link to them.

Author: I chose wikipedia, I think the tailscale link makes less sense here

The basic idea is that two endpoints, both behind a NAT, establish a
connection via a relay server.  Once the connection is established
they can do two things:

- Exchange QUIC datagrams via the relay connection.
- Coordinate holepunching to establish a direct connection.

And once you have holepunched, you can move the QUIC datagrams to the
direct connection and stop relying on the relay server.  Simple.

Note:

  This post is generally going to simplify the world a lot.  Of course
  there are many more network situations other than two endpoints both
  connected to the internet via a NAT router.  And iroh has to work
  with all of them.  But you would get bored reading this and I would
  get lost writing it.  So I'm keeping this narrative simple.

# Relay Servers

> An iroh relay server is a classical piece of server software, running

Author: Not sure I agree.  To me "a classical webserver" serves files from a directory.

> in a datacenter.  It exists even thouh we want p2p connections,

Reviewer 2: typo

  holepunched!".  Typically this will be only on 1 IP path out of all
  the candidates.  Or maybe more and more these days it'll succeed for
  both an IPv4 and and IPv6 path.
  
If you followed arefully you'll have counted 3 special messages that
need to be sent to the peer endpoint:

1. IP address candidates.  These are sent via the relay server.

2. Pings.  These are sent on the non-relayed IP paths.

3. Pongs.  These are also sent on the non-relayed IP paths.

They need to be sent as UDP datagrams.  Over the same *paths* as the
QUIC datagrams are also being sent: the **relay path** and any
**direct paths**.

# Multiplexing UDP datagrams

Iroh stands on the shoulders of giants, and it looked carefully at
ZeroTier and Tailscale.  In particular it borrowed a lot from the DERP
design from Tailscale.  From the above holepunching description we get
two kinds of packets:

- Application payload.  For iroh these are QUIC datagrams.
- Holepunching datagrams.

When an iroh endpoint receives a packet it needs to first figure out
which kind of packet this is: a QUIC datagram, or a DERP datagram?  If
it is a QUIC packet it is passed onto the QUIC stack [^quicstack].  If

> it is a DERP datagram it needs to be handed by iroh itself, by a

Reviewer: Maybe we call them "holepunching datagram" instead of "DERP datagram", to match the list above.

component we call the *magic socket*.  This is done using the "QUIC
bit", a bit in the UDP datagram defined as always set to 1 in QUIC
version 1 [^greasing].

> [^quicstack]: iroh uses Quinn for the QUIC stack, an excellet project.

Reviewer: typo "excellet"

> [^greasing]: Since QUIC has released RFC 9287 which advocateds

Reviewer: advocates

    "greasing" this bit: effectively toggeling it randomly.  This is
    an attempt to stop middleboxes from ossifying the protocol by
    starting to recognise this bit.  Iroh not being able to grease
    this bit right now is not ideal either.

# IP Congestion Control

This system works great and is what powers iroh today.  However it
also has its limitations.  One interesting aspect of the internet is
*congestion control*.  Basically IP packets get send around the
internet from router to router, and each hop has it's own speed and
capacity.  If you send too many packets the pipes will clog up and
start to slow down.  If you send yet more packets routers will start
dropping them.

Congestion control is tasked with threading the fine line of sending
as many packets as fast as possible between two endpoints, without
adversally affecting the latency and packet loss.  This is difficult
because there are many independent endpoints using all those links
between routers at the same time.  But it also has had a few decades

> of research by now, so we achieve reasonably decent results by now.

Reviewer: ```suggestion
of research, so we achieve reasonably decent results by now.
```

moves through the network and paths appear and disappear [^irohmove].

There are **a lot** of details about QUIC-MULTIPATH on how to make it
work.  And adding this functionality to Quinn has been a major
undertaking.  But the branch is becoming functional at last.

[^mpcongestion]: But hey!  Some of these paths share at least the
    first and last hop.  So they are not independent!  Indeed, they
    are not.  Congestion controllers is still a research area,
    especially for multiple paths with shared bottlenecks.  Though You
    should note that this already happens a lot on the internet, your
    laptop or phone probably has many TCP and/or QUIC connections to
    several servers right now.  And these definitely share hops.  Yet
    the congestion controllers do somehow figure out how make this
    work, at least to some degree.
    
[^irohmove]: Wait, doesn't iroh already say it can do this?  Indeed,
    indeed.  Though if you've tried this you'd have noticed your
    application did experience some hicckups for a few seconds as iroh
    was figuring out where traffic needs to go.  In theory we can do
    better with multipath, though it'll take some tweaking and

>     turning.

Reviewer: ```suggestion
    tuning.
```

So the next step we would like to take with the iroh multipath project
is to move holepunching logic itself into QUIC.  We're also not the
first to consider this: Marten Seemann and Christian Huitema have been
thinking about this as well and wrote down [some thoughts in a blog
post].  More importantly they started [QUIC-NAT-TRAVERSAL] draft which
conceptually does a simple thing: move the holepunching packets *into*
QUIC packets.

[some thoughts in a blog post]: https://seemann.io/posts/2024-10-26---p2p-quic/
[QUIC-NAT-TRAVERSAL]: https://datatracker.ietf.org/doc/draft-seemann-quic-nat-traversal/

While QUIC-NAT-TRAVERSAL is highly experimental and we don't expect to
follow it exactly as of the time of writing, this does have a number
of benefits:

- The QUIC packets are already encrypted, we no longer need to manage
  our own encryption layer separately.

- QUIC already has very advanced packet acknowledgement and loss
  recovery mechanisms.  Including the congestion control mechanisms.
  Essentially QUIC is a reliable transport, which this gets to benefit
  from.

- QUIC already has robust protection against sending too much data to
  unspececting hosts on the internet.

> - In combination with QUIC-MULTIPATH we get a very robust and flexible ys

Reviewer: The sentence here cuts off
